### ISSUE TITLE
windows configure / make

### ISSUE THREAD

**Issue by ralfixx** [petasis/tkdnd#18]
Two patches for Windows build via configure / make using Visual Studio compiler.

First is a patch to Makefile.in to recognize the .cpp extension used on windows.

Second patch is a fix for error reporting on windows: by using TCL_STATIC for Tcl_SetResult(), the char*-pointer is supposed to last a while, at least up to the next Tcl_Eval (see docs).  This is not the case, since the char[] immediately goes out-of-scope when the TKDND_REPORT_ERROR macro exits.
The patch simply uses Tcl_NewStringObj(), since the messages need to get copied anyway.

[tkdnd-patches.txt](https://github.com/petasis/tkdnd/files/2039428/tkdnd-patches.txt)

**Comment by petasis** [comment]
Dear ralfixx,

Thank you for the patch. I have applied it to the sources.
Out of curiosity, how do you build tkdnd under windows, using the visual studio compiler with configure/make?

**Comment by ralfixx** [comment]
Thanks for the swift response, Georgios!
On Windows, I have msys installed, but without the Mingw compilers. 
I start the Visual Studio Developer Command Prompt from the start menu (vcvars-something.bat) to set up the visual studio environment (PATH. INCLUDE, LIBS etc), and from that DOS-shell I invoke msys-bash as login shell.  In there, it's the usual configure/make combo.

**Comment by petasis** [comment]
I have made some changes in the git repository. Is it working as expected now?

**Comment by ralfixx** [comment]
Not quite yet:
- On Linux, the tcl-conf guesses the wrong library directory ".../lib".  I'm on 64-bit Linux, so the tclConfig.sh is located in …/lib64.  Not sure how to handle this in tcl-conf.

But even if I call configure manually, a change in Makefile.in which excludes the TK_INCLUDES from compiler flags leads to a compilation error "can't find tk.h:
   sh ./configure --with-tclinclude=../tcl8.5.19/generic --with-tkinclude=../tk8.5.19/generic …
   …
   gcc ...  -I"../tcl8.5.19/generic"  ...  -c `echo ./unix/TkDND_XDND.c` -o TkDND_XDND.o
    ./unix/TkDND_XDND.c:39:16: fatal error: tk.h: No such file or directory
     #include "tk.h"
 
As you can see in the gcc commandline, the -I.../tcl is there, but the -I.../tk is missing.
If I re-enable #148 in Makefile.in, 

```
-INCLUDES       = @PKG_INCLUDES@ @TCL_INCLUDES@
-#INCLUDES      = @PKG_INCLUDES@ @TCL_INCLUDES@ @TK_INCLUDES@ @TK_XINCLUDES@
+# INCLUDES     = @PKG_INCLUDES@ @TCL_INCLUDES@
+INCLUDES       = @PKG_INCLUDES@ @TCL_INCLUDES@ @TK_INCLUDES@ @TK_XINCLUDES@

```

compilation works again on Linux.

Windows check follows.

PS. How do you literally include text in this stupid comment function? Do I really need to include transcripts as attachments?

**Comment by ralfixx** [comment]
Same problem on windows, tk.h not found, same fix as above: revert the change in Makefile.in.

**Comment by petasis** [comment]
Yes, my fault. I updated the makefile.in to the latest Makefile.in from the sample project (to match the latest TEA), and this change slipped.
Is it working now?

**Comment by petasis** [comment]
I have committed also a fix for tcl-conf. Does it use an absolute path now?

**Comment by petasis** [comment]
Is this issue resolved?

**Comment by ralfixx** [comment]
Part 2 of the issue is resolved (fix for Makefile.in, manually calling configure and make works now).
Part 1 of the issue is not yet resolved:
tcl-conf does not find the tclConfig.sh when compiling on a 64-bit system:
- tcl-conf uses [info lib] to find the library location, this returns /path-to-tcl-installation/lib/ (location of tcl code). However, on 64-bit systems tclConfig.sh is installed in /path-to-tcl-installation/lib64/

So this results in 
```

TkDND sources: /homes/ralf/si++aegis/si++.4.0.C1065/src/tcltk/tkdnd-master
+++ Running 'bash configure -srcdir=/homes/ralf/si++aegis/si++.4.0.C1065/src/tcltk/tkdnd-master
                  --prefix=/homes/ralf/si++aegis/si++.4.0.C1065/src/tcltk/tkdnd-master/cmake/runtime
                  --exec-prefix=/homes/ralf/si++aegis/si++.4.0.C1065/src/tcltk/tkdnd-master/cmake/runtime
                  --with-tcl=/homes/ralf/si++aegis/si++.4.0.C1065/tcltk/linux/lib
                  --with-tk=/homes/ralf/si++aegis/si++.4.0.C1065/tcltk/linux/lib
                  --with-tclinclude=/homes/ralf/si++aegis/si++.4.0.C1065/tcltk/linux/include
                  --with-tkinclude=/homes/ralf/si++aegis/si++.4.0.C1065/tcltk/linux/include
                  --disable-64bit'
             in directory /homes/ralf/si++aegis/si++.4.0.C1065/src/tcltk/tkdnd-master...
configure: loading site script /usr/share/site/x86_64-unknown-linux-gnu
checking TEA configuration... ok (TEA 3.13)
configure: configuring tkdnd 2.9
checking system version... Linux-4.4.138-59-default
checking for Tcl configuration... configure: error: /homes/ralf/si++aegis/si++.4.0.C1065/tcltk/linux/lib directory doesn't contain tclConfig.sh

```

Directory structure is:

```

$ ls -l 
total 0
drwxr-xr-x ... bin
drwxr-xr-x ... include
drwxr-xr-x ... lib
drwxr-xr-x ... lib64
drwxr-xr-x ... share

$ ls -l lib
total 0
drwxr-xr-x ... tcl8
drwxr-xr-x ... tcl8.5

$ ls -l lib64
-rw-r--r-- ... libtcl8.5.so
-rw-r--r-- ... libtclstub8.5.a
drwxr-xr-x ... pkgconfig
-rw-r--r-- ... tclConfig.sh

```

I'm not sure how one can find the lib64 path from tcl code, so I can't offer a solution here.  However, since compiling via configure/make works ok, I think this is a minor issue.  Most people compiling from source know what to do, and the recommended way is cmake, anyway :-)

**Comment by ralfixx** [comment]
Plus I see right now that tcl-conf adds --disable-64bit which is also wrong on 64-bt systems...

**Comment by ralfixx** [comment]
Here is a BFI-patch which resolves both issues:
- add x86_64 to the 64-bit patterns
- if 64-bit is detected, append "64" to the libpath
With these changes, "tclsh tcl-conf" produces a working Makefile here.

```
diff -u tcl-conf\~ tcl-conf
--- tcl-conf~	2018-06-26 20:41:28.000000000 +0200
+++ tcl-conf	2018-06-29 10:09:52.533036489 +0200
@@ -13,7 +13,11 @@
 set bits --disable-64bit
 catch {
   switch $::tcl_platform(machine) {
-    amd64   {set bits --enable-64bit}
+    amd64   -
+    x86_64  {
+	set bits --enable-64bit
+	append libpath 64
+    }
     default {set bits --disable-64bit}
   }
 }

Diff finished.  Fri Jun 29 10:10:21 2018

```

**Comment by petasis** [comment]
Why is this "64" required to be added to the library path? What version of linux are you using? 
It sees it has split the tcl library scripts from the shared libraries, which is not common. I think that this should go better as a bug in TEA. Can you please post what "uname -a" returns?

**Comment by ralfixx** [comment]
Opensuse Leap 42.3, latest patchlevel.
```

uname -a
Linux panther 4.4.138-59-default #1 SMP Mon Jun 18 13:48:42 UTC 2018 (f0b8f6b) x86_64 x86_64 x86_64 GNU/Linux
```

I juts pulled a fresh copy of tcl8.6.8, did a 
```
configure --prefix=/tmp/tcl
make
make install
```

which resulted in the library structure I posted earlier:

```
% cd /tmp/tcl
% ls -l
drwxr-xr-x ... bin/
drwxr-xr-x ... include/
drwxr-xr-x ... lib/
drwxr-xr-x ... lib64/
drwxr-xr-x ... man/
drwxr-xr-x ... share/

% ls -l lib
drwxr-xr-x ... tcl8/
drwxr-xr-x ... tcl8.6/

% ls -l lib64
drwxr-xr-x ... itcl4.1.1/
-r-xr-xr-x ... libtcl8.6.so*
-rwxr-xr-x ... libtclstub8.6.a*
drwxr-xr-x ... pkgconfig/
drwxr-xr-x ... sqlite3.21.0/
drwxr-xr-x ... tcl8/
-rw-r--r-- ... tclConfig.sh
-rw-r--r-- ... tclooConfig.sh
drwxr-xr-x ... tdbc1.0.6/
drwxr-xr-x ... tdbcmysql1.0.6/
drwxr-xr-x ... tdbcodbc1.0.6/
drwxr-xr-x ... tdbcpostgres1.0.6/
drwxr-xr-x ... thread2.8.2/

```

I think this is intentional to enable 32-bit and 64-bit installations in parallel on the same 64-bit machine.

**Comment by petasis** [comment]
I have committed a "smarter" tcl-conf, which tries to locate the config files, if are not in their standard locations. Does it work in your system?

**Comment by ralfixx** [comment]
Yes, works like a charm, even with my not-so-standard location of the already-installed Tcl.

```
% tclsh tcl-conf
TkDND sources: /tmp/src/tcltk/tkdnd-master
Tcllib's package "fileutil" not found! Simulating...

Found "tclConfig.sh" in "/tmp/tcltk/linux/lib64".
Found "tkConfig.sh"  in "/tmp/tcltk/linux/lib64".

+++ Running 'bash configure -srcdir=/tmp/src/tcltk/tkdnd-master
                  --prefix=/tmp/src/tcltk/tkdnd-master/cmake/runtime
                  --exec-prefix=/tmp/src/tcltk/tkdnd-master/cmake/runtime
                  --with-tcl=/tmp/tcltk/linux/lib64
                  --with-tk=/tmp/tcltk/linux/lib64
                  --with-tclinclude=/tmp/tcltk/linux/include
                  --with-tkinclude=/tmp/tcltk/linux/include
                  --enable-64bit'
             in directory /tmp/src/tcltk/tkdnd-master...
configure: loading site script /usr/share/site/x86_64-unknown-linux-gnu
checking TEA configuration... ok (TEA 3.13)
...
config.status: creating Makefile
config.status: creating pkgIndex.tcl

```

**Comment by petasis** [comment]
Ok, then we can close this bug :-)
